src/main/java/org/mariadb/jdbc/internal/failover/impl/AuroraListener.java

```diff
@@ -154,47 +205,238 @@ public void reconnectFailedConnection(SearchFilter searchFilter) throws QueryExc
+     * @throws QueryException
+     */
+    private List<String> getCurrentEndpointIdentifiers(Protocol protocol, String urlEndStr) throws QueryException {
+        List<String> endpoints = new ArrayList<>();
+        try {
+            proxy.lock.lock();
+            try {
+                // Deleted instance may remain in db for 24 hours so ignoring instances that have had no change for IGNORE_TIME_IN_MINUTES
+                Date date = new Date();
+                TimeZone.setDefault(TimeZone.getTimeZone(protocol.getServerData("system_time_zone")));
+                int IGNORE_TIME_IN_MINUTES = 3;
+                Timestamp currentTime = new Timestamp(date.getTime() - IGNORE_TIME_IN_MINUTES*60*1000);
+
+                SingleExecutionResult queryResult = new SingleExecutionResult(null, 0, true, false);
+                protocol.executeQuery(queryResult,
+                        "select server_id, session_id from " + dbName + ".replica_host_status " +
+                                "where last_update_timestamp > '" + currentTime + "'",
+                        ResultSet.TYPE_FORWARD_ONLY);
+                MariaSelectResultSet resultSet = queryResult.getResult();
+
+                while (resultSet.next()) {
+                    endpoints.add(resultSet.getString(1) + urlEndStr);
+                }
+
+            } finally {
+                proxy.lock.unlock();
+            }
+        } catch (SQLException se) {
+            log.log(Level.WARNING, "SQL exception occurred: " + se);
+        } catch (QueryException qe) {
+            if (protocol.getProxy().hasToHandleFailover(qe)) {
+                if (masterProtocol.equals(protocol)) {
+                    setMasterHostFail();
+                } else if (secondaryProtocol.equals(protocol)) {
+                    setSecondaryHostFail();
+                }
+                addToBlacklist(protocol.getHostAddress());
+                reconnectFailedConnection(new SearchFilter(isMasterHostFail(), isSecondaryHostFail()));
+            }
+        }
+
+        return endpoints;
+    }
+
+    /**
+     * Sets urlParser if there are any changes in the instance endpoints available.
+     *
+     * @param endpoints     instance identifiers
+     * @param clusterHost   host string of the cluster
+     * @param port          port that is common to all endpoints
+     */
+    private void setUrlParserFromEndpoints(List<String> endpoints, String clusterHost, int port) {
+        List<HostAddress> addresses = urlParser.getHostAddresses();
+
+        List<String> currentHosts = new ArrayList<>();
+        for (HostAddress address: addresses) {
+            currentHosts.add(address.host);
+        }
+
+        for (int i = 0; i < addresses.size() && endpoints.size() > 0; i++) {
+            if (!addresses.get(i).host.equals(clusterHost) && !endpoints.contains(addresses.get(i).host)) {
+                removeFromBlacklist(addresses.get(i));
+                addresses.remove(i);
```

Comment:
This is just an `ArrayList` with no locking.  I am concerned about thread safety here.  I don't see anywhere else this collection is modified, but there seems to be risk in other threads not seeing the updates.